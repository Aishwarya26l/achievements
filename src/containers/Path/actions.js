export const PATH_OPEN = "PATH_OPEN";
export const pathOpen = pathId => ({
  type: PATH_OPEN,
  pathId
});

// In fact, it only fetches existance of solution, should be renamed?
export const PATH_FETCH_PROBLEMS_SOLUTIONS_SUCCESS =
  "PATH_FETCH_PROBLEMS_SOLUTIONS_SUCCESS";
export const pathFetchProblemsSolutionsSuccess = (pathId, solutions) => ({
  type: PATH_FETCH_PROBLEMS_SOLUTIONS_SUCCESS,
  pathId,
  solutions
});

export const PATH_TOGGLE_JOIN_STATUS_REQUEST =
  "PATH_TOGGLE_JOIN_STATUS_REQUEST";
export const pathToggleJoinStatusRequest = (userId, pathId, status) => ({
  type: PATH_TOGGLE_JOIN_STATUS_REQUEST,
  userId,
  pathId,
  status
});

export const PATH_TOGGLE_JOIN_STATUS_SUCCESS =
  "PATH_TOGGLE_JOIN_STATUS_SUCCESS";
export const pathToggleJoinStatusSuccess = (pathId, status) => ({
  type: PATH_TOGGLE_JOIN_STATUS_SUCCESS,
  pathId,
  status
});

export const PATH_TOGGLE_JOIN_STATUS_FAIL = "PATH_TOGGLE_JOIN_STATUS_FAIL";
export const pathToggleJoinStatusFail = (pathId, status, reason) => ({
  type: PATH_TOGGLE_JOIN_STATUS_FAIL,
  pathId,
  status,
  reason
});

export const PATH_PROBLEM_OPEN = "PATH_PROBLEM_OPEN";
export const pathProblemOpen = (pathId, problemId) => ({
  type: PATH_PROBLEM_OPEN,
  pathId,
  problemId
});
