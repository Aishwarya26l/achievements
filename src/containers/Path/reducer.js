export const path = (
  state = {
    joinedPaths: {}
  },
  action
) => {
  switch (action.type) {
    default:
      return state;
  }
};
