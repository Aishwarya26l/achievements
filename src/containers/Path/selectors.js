import { createSelector } from "reselect";

export const PATH_STATUS_OWNER = "owner";
export const PATH_STATUS_JOINED = "joined";
export const PATH_STATUS_NOT_JOINED = "not_joined";

const getUserId = state => state.firebase.auth.uid;

const getPath = (state, ownProps) =>
  ownProps.match.params.pathId[0] === "-"
    ? state.firebase.data.paths &&
      state.firebase.data.paths[ownProps.match.params.pathId] && {
        ...state.firebase.data.paths[ownProps.match.params.pathId],
        id: ownProps.match.params.pathId
      }
    : {
        id: ownProps.match.params.pathId,
        owner: ownProps.match.params.pathId,
        name: "Default"
      };

const getJoinedPaths = state => state.paths.joinedPaths || {};

const getProblems = (state, ownProps) => {
  const path = getPath(state, ownProps);

  return (
    path &&
    path.owner &&
    state.firebase.data.problems &&
    state.firebase.data.problems[path.owner]
  );
};

export const pathStatusSelector = createSelector(
  getUserId,
  getJoinedPaths,
  getPath,
  (userId, joinedPaths, path) => {
    if (!path || path.owner === userId) {
      return PATH_STATUS_OWNER;
    }
    if (joinedPaths[path.id]) {
      return PATH_STATUS_JOINED;
    }
    return PATH_STATUS_NOT_JOINED;
  }
);

export const pathProblemsSelector = createSelector(
  getPath,
  getProblems,
  (path, problems) => ({
    path: path,
    problems: Object.keys(problems || {})
      .map(id => ({ ...problems[id], id }))
      .filter(
        problem =>
          path.id === path.owner ? !problem.path : problem.path === path.id
      )
  })
);
