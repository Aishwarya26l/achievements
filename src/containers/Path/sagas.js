import { call, put, select, take, takeLatest } from "redux-saga/effects";
import { PATH_OPEN, pathFetchProblemsSolutionsSuccess } from "./actions";
import { pathsService } from "../../services/paths";
import { PATHS_JOINED_FETCH_SUCCESS } from "../Paths/actions";

export function* pathOpenHandler(action) {
  if (!action.pathId || action.pathId[0] !== "-") {
    return yield Promise.resolve();
  }

  let owner = yield select(
    state =>
      state.paths &&
      state.paths.joinedPaths &&
      state.paths.joinedPaths[action.pathId] &&
      state.paths.joinedPaths[action.pathId].owner
  );
  if (!owner) {
    yield take(PATHS_JOINED_FETCH_SUCCESS);
    owner = yield select(
      state =>
        state.paths &&
        state.paths.joinedPaths &&
        state.paths.joinedPaths[action.pathId] &&
        state.paths.joinedPaths[action.pathId].owner
    );
  }
  // FIXIT: join in previous select
  const uid = yield select(state => state.firebase.auth.uid);
  const problems = yield call(
    [pathsService, pathsService.fetchProblems],
    owner,
    action.pathId
  );

  const solutions = yield call(
    [pathsService, pathsService.fetchProblemsSolutions],
    uid,
    problems
  );

  yield put(pathFetchProblemsSolutionsSuccess(action.pathId, solutions));
}

export default [
  function* watchPathOpenRequest() {
    yield takeLatest(PATH_OPEN, pathOpenHandler);
  }
];
