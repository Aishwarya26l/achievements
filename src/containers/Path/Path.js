/**
 * @file Path container module
 * @author Theodor Shaytanov <theodor.shaytanov@gmail.com>
 * @created 17.03.18
 */
import React, { Fragment } from "react";
import PropTypes from "prop-types";
import { compose } from "redux";
import { connect } from "react-redux";

import { APP_SETTING } from "../../achievementsApp/config";

import Button from "@material-ui/core/Button";
import Toolbar from "@material-ui/core/Toolbar";

import ProblemsTable from "../../components/tables/ProblemsTable";

import { firebaseConnect } from "react-redux-firebase";

import withRouter from "react-router-dom/withRouter";
import Breadcrumbs from "../../components/Breadcrumbs";
import {
  pathStatusSelector,
  pathProblemsSelector,
  PATH_STATUS_JOINED,
  PATH_STATUS_OWNER,
  PATH_STATUS_NOT_JOINED
} from "./selectors";
import { pathOpen, pathToggleJoinStatusRequest } from "./actions";
import { pathProblemDialogShow } from "../Paths/actions";
import ProblemDialog from "../../components/dialogs/ProblemDialog";

import AddIcon from "@material-ui/icons/Add";
import { sagaInjector } from "../../services/saga";
import sagas from "./sagas";

class Path extends React.PureComponent {
  static propTypes = {
    dispatch: PropTypes.func,
    match: PropTypes.object,
    pathProblems: PropTypes.object,
    pathStatus: PropTypes.string,
    ui: PropTypes.any,
    uid: PropTypes.string
  };

  componentDidMount() {
    this.props.dispatch(pathOpen(this.props.match.params.pathId));
  }

  changeJoinStatus = () =>
    this.props.dispatch(
      pathToggleJoinStatusRequest(
        this.props.uid,
        this.props.pathProblems.path.id,
        this.props.pathStatus === PATH_STATUS_NOT_JOINED
      )
    );

  onAddProblemClick = () => this.props.dispatch(pathProblemDialogShow());

  render() {
    const { dispatch, match, pathProblems, pathStatus, ui, uid } = this.props;

    let pathName;

    if (match.params.pathId[0] !== "-") {
      pathName = "Default";
    }

    pathName = pathName || (pathProblems.path && pathProblems.path.name) || "";

    return (
      <Fragment>
        <Breadcrumbs
          action={
            pathStatus === PATH_STATUS_OWNER
              ? null
              : {
                  label: pathStatus === PATH_STATUS_JOINED ? "Leave" : "Join",
                  handler: this.changeJoinStatus.bind(this)
                }
          }
          paths={[
            {
              label: "Paths",
              link: "/paths"
            },
            {
              label: pathName
            }
          ]}
        />
        {pathStatus === PATH_STATUS_OWNER &&
          (!APP_SETTING.isSuggesting ? (
            <Toolbar>
              <Button
                color="primary"
                onClick={this.onAddProblemClick}
                variant="raised"
              >
                Add Problem
              </Button>
            </Toolbar>
          ) : (
            <Button
              aria-label="Add"
              color="primary"
              onClick={this.onAddProblemClick}
              style={{
                position: "fixed",
                bottom: 20,
                right: 20
              }}
              variant="fab"
            >
              <AddIcon />
            </Button>
          ))}
        <ProblemsTable
          currentUserId={uid}
          dispatch={dispatch}
          pathOwnerId={pathProblems.path && pathProblems.path.owner}
          problems={pathProblems.problems || []}
          selectedPathId={(pathProblems.path && pathProblems.path.id) || ""}
        />
        <ProblemDialog
          dispatch={dispatch}
          open={ui.dialog.type === "ProblemChange"}
          pathId={(pathProblems.path && pathProblems.path.id) || ""}
          problem={ui.dialog.value}
        />
      </Fragment>
    );
  }
}

sagaInjector.inject(sagas);

const mapStateToProps = (state, ownProps) => ({
  pathProblems: pathProblemsSelector(state, ownProps),
  pathStatus: pathStatusSelector(state, ownProps),
  ui: state.paths.ui,
  uid: state.firebase.auth.uid
});

export default compose(
  withRouter,
  firebaseConnect((ownProps, store) => {
    const state = store.getState();
    const uid = state.firebase.auth.uid;
    const pathId = ownProps.match.params.pathId;

    if (!uid) {
      return false;
    }

    return [
      `/paths/${pathId}`,
      "/problems",
      `/studentJoinedPaths/${uid}/${pathId}`
    ];
  }),
  connect(mapStateToProps)
)(Path);
