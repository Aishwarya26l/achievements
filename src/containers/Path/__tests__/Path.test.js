import React from "react";
import { createShallow } from "@material-ui/core/test-utils";
import { Path } from "../Path";

// noinspection RequiredAttributes
describe("<Path>", () => {
  let shallow;
  let spy;

  beforeEach(() => {
    shallow = createShallow();
    spy = jest.fn();
  });

  it("should check snapshot", () => {
    const wrapper = shallow(
      <Path
        match={{
          params: {
            pathId: "testPath"
          }
        }}
        onAddAssistant={spy}
        onAssistantKeyChange={spy}
        onCloseDialog={spy}
        onOpen={spy}
        onProblemChangeRequest={spy}
        onRemoveAssistant={spy}
        pathActivities={{
          activities: [],
          path: {
            id: ""
          }
        }}
        pathStatus="test"
        ui={{
          dialog: false
        }}
        uid="testUser"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
