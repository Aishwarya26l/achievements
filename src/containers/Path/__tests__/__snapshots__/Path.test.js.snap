// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Path> should check snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Path
    match={
        Object {
            "params": Object {
              "pathId": "testPath",
            },
          }
    }
    onAddAssistant={[Function]}
    onAssistantKeyChange={[Function]}
    onCloseDialog={[Function]}
    onOpen={[Function]}
    onProblemChangeRequest={[Function]}
    onRemoveAssistant={[Function]}
    pathActivities={
        Object {
            "activities": Array [],
            "path": Object {
              "id": "",
            },
          }
    }
    pathStatus="test"
    ui={
        Object {
            "dialog": false,
          }
    }
    uid="testUser"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <WithStyles(Breadcrumbs)
          action={
                    Array [
                              Object {
                                "handler": [Function],
                                "label": "Request more",
                              },
                              false,
                              Object {
                                "handler": [Function],
                                "label": "Join",
                              },
                            ]
          }
          paths={
                    Array [
                              Object {
                                "label": "Paths",
                                "link": "/paths",
                              },
                              Object {
                                "label": "Default",
                              },
                            ]
          }
/>,
        false,
        <AddTextSolutionDialog
          onClose={[Function]}
          onCommit={[Function]}
          open={false}
          solution={undefined}
          taskId={undefined}
/>,
        <WithStyles(AddJestSolutionDialog)
          onClose={[Function]}
          onCommit={[Function]}
          open={false}
          problem={undefined}
          taskId={undefined}
/>,
        <AddProfileDialog
          defaultValue=""
          externalProfile={
                    Object {
                              "description": "learn to Code JavaScript by Playing a Game",
                              "id": "CodeCombat",
                              "name": "Code Combat",
                              "url": "https://codecombat.com",
                            }
          }
          onClose={[Function]}
          onCommit={[Function]}
          open={false}
/>,
        <WithStyles(ActivitiesTable)
          activities={Array []}
          currentUserId="testUser"
          onEditProblem={undefined}
          onMoveProblem={[Function]}
          onOpenProblem={[Function]}
          pathOwnerId={undefined}
          selectedPathId=""
/>,
        <ActivityDialog
          activity={undefined}
          onClose={[Function]}
          onCommit={[Function]}
          open={false}
          pathId=""
          uid="testUser"
/>,
        <ControlAssistantsDialog
          assistants={false}
          newAssistant={false}
          onAddAssistant={[Function]}
          onAssistantKeyChange={[Function]}
          onClose={[Function]}
          onRemoveAssistant={[Function]}
          open={false}
          target=""
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "action": Array [
            Object {
              "handler": [Function],
              "label": "Request more",
            },
            false,
            Object {
              "handler": [Function],
              "label": "Join",
            },
          ],
          "paths": Array [
            Object {
              "label": "Paths",
              "link": "/paths",
            },
            Object {
              "label": "Default",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onClose": [Function],
          "onCommit": [Function],
          "open": false,
          "solution": undefined,
          "taskId": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onClose": [Function],
          "onCommit": [Function],
          "open": false,
          "problem": undefined,
          "taskId": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "defaultValue": "",
          "externalProfile": Object {
            "description": "learn to Code JavaScript by Playing a Game",
            "id": "CodeCombat",
            "name": "Code Combat",
            "url": "https://codecombat.com",
          },
          "onClose": [Function],
          "onCommit": [Function],
          "open": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "activities": Array [],
          "currentUserId": "testUser",
          "onEditProblem": undefined,
          "onMoveProblem": [Function],
          "onOpenProblem": [Function],
          "pathOwnerId": undefined,
          "selectedPathId": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "activity": undefined,
          "onClose": [Function],
          "onCommit": [Function],
          "open": false,
          "pathId": "",
          "uid": "testUser",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "assistants": false,
          "newAssistant": false,
          "onAddAssistant": [Function],
          "onAssistantKeyChange": [Function],
          "onClose": [Function],
          "onRemoveAssistant": [Function],
          "open": false,
          "target": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <WithStyles(Breadcrumbs)
            action={
                        Array [
                                    Object {
                                      "handler": [Function],
                                      "label": "Request more",
                                    },
                                    false,
                                    Object {
                                      "handler": [Function],
                                      "label": "Join",
                                    },
                                  ]
            }
            paths={
                        Array [
                                    Object {
                                      "label": "Paths",
                                      "link": "/paths",
                                    },
                                    Object {
                                      "label": "Default",
                                    },
                                  ]
            }
/>,
          false,
          <AddTextSolutionDialog
            onClose={[Function]}
            onCommit={[Function]}
            open={false}
            solution={undefined}
            taskId={undefined}
/>,
          <WithStyles(AddJestSolutionDialog)
            onClose={[Function]}
            onCommit={[Function]}
            open={false}
            problem={undefined}
            taskId={undefined}
/>,
          <AddProfileDialog
            defaultValue=""
            externalProfile={
                        Object {
                                    "description": "learn to Code JavaScript by Playing a Game",
                                    "id": "CodeCombat",
                                    "name": "Code Combat",
                                    "url": "https://codecombat.com",
                                  }
            }
            onClose={[Function]}
            onCommit={[Function]}
            open={false}
/>,
          <WithStyles(ActivitiesTable)
            activities={Array []}
            currentUserId="testUser"
            onEditProblem={undefined}
            onMoveProblem={[Function]}
            onOpenProblem={[Function]}
            pathOwnerId={undefined}
            selectedPathId=""
/>,
          <ActivityDialog
            activity={undefined}
            onClose={[Function]}
            onCommit={[Function]}
            open={false}
            pathId=""
            uid="testUser"
/>,
          <ControlAssistantsDialog
            assistants={false}
            newAssistant={false}
            onAddAssistant={[Function]}
            onAssistantKeyChange={[Function]}
            onClose={[Function]}
            onRemoveAssistant={[Function]}
            open={false}
            target=""
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "action": Array [
              Object {
                "handler": [Function],
                "label": "Request more",
              },
              false,
              Object {
                "handler": [Function],
                "label": "Join",
              },
            ],
            "paths": Array [
              Object {
                "label": "Paths",
                "link": "/paths",
              },
              Object {
                "label": "Default",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onClose": [Function],
            "onCommit": [Function],
            "open": false,
            "solution": undefined,
            "taskId": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onClose": [Function],
            "onCommit": [Function],
            "open": false,
            "problem": undefined,
            "taskId": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "defaultValue": "",
            "externalProfile": Object {
              "description": "learn to Code JavaScript by Playing a Game",
              "id": "CodeCombat",
              "name": "Code Combat",
              "url": "https://codecombat.com",
            },
            "onClose": [Function],
            "onCommit": [Function],
            "open": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activities": Array [],
            "currentUserId": "testUser",
            "onEditProblem": undefined,
            "onMoveProblem": [Function],
            "onOpenProblem": [Function],
            "pathOwnerId": undefined,
            "selectedPathId": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activity": undefined,
            "onClose": [Function],
            "onCommit": [Function],
            "open": false,
            "pathId": "",
            "uid": "testUser",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "assistants": false,
            "newAssistant": false,
            "onAddAssistant": [Function],
            "onAssistantKeyChange": [Function],
            "onClose": [Function],
            "onRemoveAssistant": [Function],
            "open": false,
            "target": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {},
  },
}
`;
